pymodbus/transaction.py

```diff
@@ -238,31 +230,37 @@ def _recv(self, expected_response_length, full):
                 min_size = expected_response_length
 
             read_min = self.client.framer.recvPacket(min_size)
-            if read_min:
+            if not read_min:
+                return read_min
+
+            if len(read_min) < min_size:
+                raise InvalidMessageReceivedException(
+                    "Incomplete message received, expected at least %d bytes (%d received)"
+                    % (min_size, len(read_min)))
+
+            if isinstance(self.client.framer, ModbusSocketFramer):
+                func_code = byte2int(read_min[-1])
+            elif isinstance(self.client.framer, ModbusRtuFramer):
+                func_code = byte2int(read_min[-1])
+            elif isinstance(self.client.framer, ModbusAsciiFramer):
+                func_code = int(read_min[3:5], 16)
+            elif isinstance(self.client.framer, ModbusBinaryFramer):
+                func_code = byte2int(read_min[-1])
+            else:
+                func_code = -1
+
+            if func_code < 0x80:    # Not an error
                 if isinstance(self.client.framer, ModbusSocketFramer):
-                    func_code = byte2int(read_min[-1])
-                elif isinstance(self.client.framer, ModbusRtuFramer):
-                    func_code = byte2int(read_min[-1])
-                elif isinstance(self.client.framer, ModbusAsciiFramer):
-                    func_code = int(read_min[3:5], 16)
-                elif isinstance(self.client.framer, ModbusBinaryFramer):
-                    func_code = byte2int(read_min[-1])
-                else:
-                    func_code = -1
-
-                if func_code < 0x80:    # Not an error
-                    if isinstance(self.client.framer, ModbusSocketFramer):
-                        # Ommit UID, which is included in header size
-                        h_size = self.client.framer._hsize
-                        length = struct.unpack(">H", read_min[4:6])[0] - 1
-                        expected_response_length = h_size + length
-                    expected_response_length -= min_size
-                    total = expected_response_length + min_size
-                else:
-                    expected_response_length = exception_length - min_size
-                    total = expected_response_length + min_size
+                    # Ommit UID, which is included in header size
+                    h_size = self.client.framer._hsize
+                    length = struct.unpack(">H", read_min[4:6])[0] - 1
+                    expected_response_length = h_size + length
+                expected_response_length -= min_size
```

Comment:
This would throw `TypeError` (`None - min_size`) for messages with out the `get_response_pdu_size` method defined (for e.g. all mei_messages and diagnostic messages). You can run `synchronous_client_ext.py` under `examples\common` to check this behaviour.